File: 2019/ruby/day_01.rb
file_path = File.expand_path("../../inputs/day_01.txt", __FILE__)
input     = File.open(file_path).readlines.map(&:to_i)

def star_01(masses)
  masses.map { |mass| calc_fuel(mass) }.sum
end

def star_02(masses)
  masses.map do |mass|
    fuel = 0
    mass = calc_fuel(mass)
    while mass > 0
      fuel += mass
      mass = calc_fuel(mass)
    end
    fuel
  end.sum
end

def calc_fuel(mass)
 (mass / 3).floor - 2
end

puts star_01(input) # 3514064
puts star_02(input) # 5268207

File: 2019/javascript/day_01.js
const fs = require("fs")
const path = require("path")
const assert = require("assert")
const file_path = path.resolve(__dirname, "../inputs/day_01.txt")
const input = fs.readFileSync(file_path, "utf8")
  .split("\n")
  .filter(Boolean)
  .map(line => Number.parseInt(line))

function star_01(masses) {
  return masses.map(calc_fuel).reduce((acc, mass) => acc + mass)
}

function star_02(masses) {
  return masses.map(mass => {
    let fuel = 0
    while ((mass = calc_fuel(mass)) > 0) {
      fuel += mass
    }
    return fuel
  }).reduce((acc, mass) => acc + mass)
}

function calc_fuel(mass) {
  return Math.floor(mass / 3) - 2
}

assert.equal(star_01(input), 3514064)
assert.equal(star_02(input), 5268207)
